Improve cycles output in `analyze` tool

Overcastan opened the pull request:
This PR adds extended information about traces lengths of VM components to the `analyze` tool output.

bobbinth commented on miden/src/tools/mod.rs:
```diff
@@ -149,19 +208,34 @@ pub fn analyze<A>(
+
     let mut execution_details = ExecutionDetails::default();
 
+    // if the `cycles_info` flag is true, get the program execution trace to obtain the trace length summary
+    if cycles_info {
+        let trace = processor::execute(
+            &program,
+            stack_inputs.clone(),
+            advice_provider.clone(),
+            ExecutionOptions::default(),
+        )
+        .map_err(|err| ProgramError::ExecutionError(err))?;
+        execution_details.set_extended_cycle_info(trace.trace_len_summary());
+    }
+
+    let vm_state_iterator = processor::execute_iter(&program, stack_inputs, advice_provider);
```
Doing `processor:execute()` will cause us to execute the program twice. Is there a way to modify `vm_state_iterator` to get that info from it?

Overcastan replied:
Hmm, first I wanted to say that we can obtain info about operation only from `execute_iter` and info about VM cycles only from trace, which we can get from `execute`, but now I think we can get around it. I can try to implement another function, something like `execute_with_iter`, from which I can obtain both iterator and execution trace. Sorry @bobbinth, my bad, I'll rework this

bobbinth replied:
You can probably just modify the [VmStateIterator](https://github.com/0xPolygonMiden/miden-vm/blob/next/processor/src/debug.rs#L50) to expose this extra property.

bobbinth commented on processor/src/lib.rs:
```diff
@@ -147,6 +147,30 @@ where
     VmStateIterator::new(process, result)
 }
 
+/// Returns a tuple with VM state iterator and trace lengths summary.
+pub fn execute_iter_with_trace_len<H>(
+    program: &Program,
+    stack_inputs: StackInputs,
+    host: H,
+) -> (VmStateIterator, TraceLenSummary)
```
I would prefer not to introduce a new method but just to add `trace_len_summary` to the `VmStateIterator` struct.

bobbinth commented on processor/src/debug.rs:
```diff
@@ -55,14 +56,23 @@ pub struct VmStateIterator {
-        let (system, decoder, stack, _, chiplets, _) = process.into_parts();
+        let (system, decoder, stack, mut range, chiplets, _) = process.into_parts();
+
+        let clk = system.clk();
+        let range_table_len = range.get_number_range_checker_rows();
+        chiplets.append_range_checks(&mut range);
+
+        let trace_len_summary =
+            TraceLenSummary::new(clk as usize, range_table_len, ChipletsLengths::new(&chiplets));
```
Let's create a separate function to encapsulate this logic. Maybe something like `build_trace_len_summary(system, range, chipletes)`.

bobbinth commented on processor/src/trace/utils.rs:
```diff
@@ -315,6 +315,20 @@ impl ChipletsLengths {
         }
     }
 
+    pub fn new_debug(
+        hash_len: usize,
+        bitwise_len: usize,
+        memory_len: usize,
+        kernel_len: usize,
+    ) -> Self {
+        ChipletsLengths {
+            hash_chiplet_len: hash_len,
+            bitwise_chiplet_len: bitwise_len,
+            memory_chiplet_len: memory_len,
+            kernel_rom_len: kernel_len,
+        }
+    }
```
I would probably rename this into `from_parts` and since we use it only for debugging purposes, I'd add `#[cfg(test)]` attribute to it.

bobbinth reviewed (approved):
Looks good! Thank you! I left a couple of minor comments inline. After these are addressed, we can merge.

Also, could you paste a screenshot of how the output for this looks like?

Overcastan commented:
Now `analyze` output for the `merkle_store` example looks like this: 

<img width="1257" alt="Screenshot 2023-10-10 at 9 52 53 PM" src="https://github.com/0xPolygonMiden/miden-vm/assets/43154072/024d3b3d-f54c-4a61-84c7-9fda7cf8c27d">

and similar output for the `fibonacci` example: 

<img width="976" alt="Screenshot 2023-10-10 at 9 55 02 PM" src="https://github.com/0xPolygonMiden/miden-vm/assets/43154072/fb9e86a1-c551-453c-9f3b-2edf3c833a37">

I added variable width for the first column, so operations names will always fit (for example, before that `push.*.*.*` operations from `merkle_store` example were breaking the table). I also added the line to highlight the table header.

bobbinth commented:
Looks great! A couple of minor comments on the output formatting:

1. Let's move "Total number of NOOPs executed" after the table.
  a. Also, let's not capitalize "number".
2. For the table, let's rename "AsmOp" title into "Assembly instruction"
3. Above the "VM cycles" section, let's add a header "Analyzed [program name] with [input name]".

Basically, the output could look like:
```
============================================================
Analyzed fib.masm program with fib.inputs

VM cycles: ...

Assembly Instruction ..
--------------------------

Total number of NOOPs executed: 
```

Overcastan commented:
After last refactoring output looks like this: 

<img width="1227" alt="Screenshot 2023-10-11 at 1 00 56 AM" src="https://github.com/0xPolygonMiden/miden-vm/assets/43154072/e141f722-9f43-4aa1-a0a9-13a70a9ead47">

<img width="981" alt="Screenshot 2023-10-11 at 1 01 30 AM" src="https://github.com/0xPolygonMiden/miden-vm/assets/43154072/3001f084-7482-4a2b-a6e4-0576306761e4">

bobbinth commented:
Looks good! Thank you! One last thing: could we add:
```
============================================================
```
Before the `Analyzed...` line?

Overcastan commented:
The final view: 

<img width="1228" alt="Screenshot 2023-10-11 at 1 23 26 AM" src="https://github.com/0xPolygonMiden/miden-vm/assets/43154072/e10f9a78-9269-4714-a47c-b0ca5954db83">

bobbinth commented:
All looks good! Thank you!
